File: src-tauri/src/events.rs
#![allow(dead_code)]

use serde::Serialize;

use crate::commands::TransportKind;

pub const ATTACK_PROGRESS: &str = "attack://progress";
pub const ATTACK_FOUND_KEY: &str = "attack://found-key";
pub const ATTACK_SUMMARY: &str = "attack://summary";
pub const ATTACK_HARDNESTED: &str = "attack://hardnested";
pub const ATTACK_ERROR: &str = "attack://error";
pub const DEVICE_STATUS: &str = "device://status";
pub const TRANSFER_PROGRESS: &str = "transfer://progress";
pub const AUTO_STATUS: &str = "auto://status";
pub const AUTO_SUMMARY: &str = "auto://summary";
pub const AUTO_ERROR: &str = "auto://error";

#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct AttackProgressPayload {
    pub stage: String,
    pub processed: u64,
    pub total: u64,
    pub percent: f32,
}

#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct FoundKeyPayload {
    pub key: String,
    pub uid: Option<String>,
    pub key_type: Option<String>,
}

#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct DictOutputPayload {
    pub uid: String,
    pub path: String,
    pub key_count: u64,
}

#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct AttackSummaryPayload {
    pub found_keys: u64,
    pub candidate_keys: u64,
    pub dict_outputs: Vec<DictOutputPayload>,
    pub status: String,
}

#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct HardNestedPayload {
    pub line: String,
}

#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct DeviceStatusPayload {
    pub status: String,
    pub device_id: Option<String>,
    pub transport: Option<TransportKind>,
    pub code: Option<String>,
    pub message: Option<String>,
}

#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct TransferProgressPayload {
    pub path: String,
    pub transferred: u64,
    pub total: u64,
    pub percent: f32,
}

#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct AutoStatusPayload {
    pub phase: String,
    pub message: Option<String>,
}

#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct AutoSummaryPayload {
    pub found_keys: u64,
    pub uploaded_dicts: u64,
    pub keys_added: u64,
    pub keys_uploaded: bool,
    pub logs_total: u64,
    pub logs_skipped: u64,
}

File: src/lib/ipc/events.ts
import { listen, type UnlistenFn } from "@tauri-apps/api/event";
import { isTauri } from "@tauri-apps/api/core";
import type {
  AttackProgress,
  DictOutput,
  FoundKey,
} from "../../store/useAttackStore";
import type {
  ConnectionStatus,
  TransportKind,
} from "../../store/useConnectionStore";
import type { AutoPhase } from "../../store/useAutoStore";
import type { CommandError } from "./commands";

export type AttackCompletionStatus = "success" | "cancelled" | "empty";

export interface AttackSummaryPayload {
  foundKeys: number;
  candidateKeys: number;
  dictOutputs: DictOutput[];
  status: AttackCompletionStatus;
}

export interface HardNestedPayload {
  line: string;
}

export interface DeviceStatusPayload {
  status: ConnectionStatus;
  deviceId?: string;
  transport?: TransportKind;
  code?: string;
  message?: string;
}

export interface TransferProgressPayload {
  path: string;
  transferred: number;
  total: number;
  percent: number;
}

export interface AutoStatusPayload {
  phase: AutoPhase;
  message?: string;
}

export interface AutoSummaryPayload {
  foundKeys: number;
  uploadedDicts: number;
  keysAdded: number;
  keysUploaded: boolean;
  logsTotal: number;
  logsSkipped: number;
}

export interface EventPayloadMap {
  "attack://progress": AttackProgress;
  "attack://found-key": FoundKey;
  "attack://summary": AttackSummaryPayload;
  "attack://hardnested": HardNestedPayload;
  "attack://error": CommandError;
  "device://status": DeviceStatusPayload;
  "transfer://progress": TransferProgressPayload;
  "auto://status": AutoStatusPayload;
  "auto://summary": AutoSummaryPayload;
  "auto://error": CommandError;
}

export type EventName = keyof EventPayloadMap;

export const EVENTS = {
  attackProgress: "attack://progress",
  attackFoundKey: "attack://found-key",
  attackSummary: "attack://summary",
  attackHardNested: "attack://hardnested",
  attackError: "attack://error",
  deviceStatus: "device://status",
  transferProgress: "transfer://progress",
  autoStatus: "auto://status",
  autoSummary: "auto://summary",
  autoError: "auto://error",
} as const;

export function onTauriEvent<K extends EventName>(
  name: K,
  handler: (payload: EventPayloadMap[K]) => void,
): Promise<UnlistenFn> {
  if (!isTauri()) {
    return Promise.resolve(() => {});
  }
  return listen<EventPayloadMap[K]>(name, (event) => handler(event.payload));
}

export function throttle<A extends unknown[]>(
  fn: (...args: A) => void,
  intervalMs: number,
): (...args: A) => void {
  let last = 0;
  let pending: A | null = null;
  let timer: ReturnType<typeof setTimeout> | null = null;

  const run = (args: A) => {
    last = Date.now();
    fn(...args);
  };

  return (...args: A) => {
    const remaining = intervalMs - (Date.now() - last);
    if (remaining <= 0) {
      if (timer) {
        clearTimeout(timer);
        timer = null;
      }
      run(args);
      return;
    }
    pending = args;
    if (!timer) {
      timer = setTimeout(() => {
        timer = null;
        if (pending) {
          run(pending);
          pending = null;
        }
      }, remaining);
    }
  };
}
